src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -236,33 +236,49 @@ public static FormulaValue Text(EvalVisitor runner, EvalVisitorContext context,
                     }
 
                     break;
-                case DateValue:
-                case DateTimeValue:
-                case TimeValue:
-                    DateTimeValue dateTimeValue;
+                case DateTimeValue dateTimeValue:
+                    if (formatString != null && hasNumberFmt)
+                    {
+                        // It's a datetime, formatted as number. Let's convert it to a number value first
+                        var newNumber = Library.DateTimeToNumber(runner, context, IRContext.NotInSource(FormulaType.Number), new DateTimeValue[] { dateTimeValue });
+                        resultString = newNumber.Value.ToString(formatString ?? "g", culture);
+                    }
+                    else
+                    {
+                        return ExpandDateTimeExcelFormatSpecifiers(irContext, formatString, "g", dateTimeValue.GetConvertedValue(timeZoneInfo), culture, runner.CancellationToken);
+                    }
 
-                    if (args[0] is TimeValue t)
+                    break;
+                case DateValue dateValue:
+                    if (formatString != null && hasNumberFmt)
                     {
-                        dateTimeValue = Library.TimeToDateTime(runner, context, IRContext.NotInSource(FormulaType.DateTime), new TimeValue[] { t });
+                        NumberValue newDateNumber = Library.DateToNumber(runner, context, IRContext.NotInSource(FormulaType.Number), new DateValue[] { dateValue }) as NumberValue;
+                        resultString = newDateNumber.Value.ToString(formatString ?? "g", culture);
                     }
-                    else if (args[0] is DateValue d)
+                    else if (formatString != null)
                     {
-                        dateTimeValue = FormulaValue.New(d.GetConvertedValue(timeZoneInfo));
+                        return ExpandDateTimeExcelFormatSpecifiers(irContext, formatString, "g", dateValue.GetConvertedValue(timeZoneInfo), culture, runner.CancellationToken);
                     }
                     else
                     {
-                        dateTimeValue = args[0] as DateTimeValue;
+                        resultString = dateValue.GetConvertedValue(timeZoneInfo).ToShortDateString();
                     }
```

Comment:
Can we instead reuse else if part and specify defaultFormat to be "d" ?
[https://learn.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=net-7.0](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=net-7.0)

Reply:
Updated. thx